reconcile/external_resources/manager.py

```diff
@@ -0,0 +1,578 @@
+                resource_type="Job",
+                name=job_name,
+            )
+        else:
+            _obj = self.oc.get(
+                namespace=self.namespace,
+                kind="Job",
+                name=job_name,
+                allow_not_found=True,
+            )
+            if _obj:
+                obj = OpenshiftResource(
+                    _obj, QONTRACT_INTEGRATION, QONTRACT_INTEGRATION_VERSION
+                )
+
+        if obj is None:
+            return ReconcileStatus.NOT_EXISTS
+
+        status = obj.body["status"]
+        backofflimit = obj.body["spec"].get("backoffLimit", 6)
+        if status.get("succeeded", 0) > 0:
+            return ReconcileStatus.SUCCESS
+        elif status.get("failed", 0) >= backofflimit:
+            return ReconcileStatus.ERROR
+        else:
+            return ReconcileStatus.IN_PROGRESS
+
+    def reconcile_resource(
+        self,
+        reconciliation: Reconciliation,
+        module: ExternalResourceModule,
+        dry_run: bool,
+    ) -> None:
+        job_name = self._get_job_name(reconciliation)
+        current_job = self.ri.get_current(self.cluster, self.namespace, "Job", job_name)
+
+        if current_job:
+            logging.info(
+                "Removing previous Job. Name: %s, Action: %s, Key: %s",
+                job_name,
+                reconciliation.action,
+                reconciliation.key,
+            )
+            self.oc.delete(self.namespace, "Job", job_name)
+
+        logging.info(
+            "Spawning Reconciliation Job. Name: %s, Action: %s, IsDryRun: %s, Key: %s",
+            job_name,
+            reconciliation.action,
+            dry_run,
+            reconciliation.key,
+        )
+        job_spec = self._build_job_spec(reconciliation=reconciliation, module=module)
+        job = OpenshiftResource(
+            job_spec, QONTRACT_INTEGRATION, QONTRACT_INTEGRATION_VERSION
+        )
+        dj = job.annotate()
+        self.oc.apply(self.namespace, dj)
+
+    def get_resource_reconcile_logs(self, reconciliation: Reconciliation) -> None:
+        name = self._get_job_name(reconciliation)
+        self.oc.job_logs(
+            namespace=self.namespace, follow=False, name=name, output=sys.stdout
+        )
+
+
+class ExternalResourcesManager:
+    def __init__(
+        self,
+        oc: OCCli,
+        cluster: str,
+        namespace: str,
+        state_manager: ExternalResourcesStateDynamoDB,
+        settings: ExternalResourcesSettings,
+        modules: Mapping[ExternalResourceModuleKey, ExternalResourceModule],
+        impl: ExternalResourcesImpl,
+        dry_run: bool = False,
+    ) -> None:
+        self.state_mgr = state_manager
+        self.oc = oc
+        self.cluster = cluster
+        self.namespace = namespace
+        self.settings = settings
+        self.modules = modules
+        self.impl = impl
+        self.dry_run = dry_run
+        self.providers = ObjectFactory[ExternalResourceFactory]()
```

Comment:
A great use case for factory pattern!
So, extending for i.e. GCP would simply mean registering a new factory?

Reply:
This changed a bit. There 2 levels of factories, 1 to use tech stacks and another one for providers. 

1.- CDKTF/Terraform, others
2.- AWS-IAM-ROLE, CLOUDFLARE-..., GCP-... 

But yes, the idea is to isolate different use-cases/tech-stacks.